### PR TITLE
feat(bitbucket): Bitbucket Server platform support

### LINKED ISSUE #2482: Support for Bitbucket Server
**What would you like Renovate to be able to do?**
I would like to use Renovate with our own hosted Bitbucket Server

**Describe the solution you'd like**
Please implement a platform backend for Bitbucket Server. I can help.

**Describe alternatives you've considered**
Currently i can't use Renovate and there is no workaround

**Additional context**
[Bitbucket Server REST API](https://docs.atlassian.com/bitbucket-server/rest/5.13.1/bitbucket-rest.html)

### PR DESCRIPTION
**Pull request by iamstarkov**
Closes #2482

### CODE REVIEW

**Comment by kayoub5** [comment]
Is it just me, or bitbucket API for cloud and server diverge too much, to the point where we should consider it a different platform?

**Comment by rarkins** [comment]
It does start to seem like they could be separate platforms. I think the current approach to the PR is good - just change everything necessary and if we decide it’s not worth “unifying” them again then we can split into two platforms with a little shared code.

**Comment by iamstarkov** [comment]
yeah, basic concepts are the same, but might have new names, also set of endpoints is different a little

**Comment by iamstarkov** [comment]
we thought we would start with existing adapter to see how big are the moving parts

**Comment by rarkins** [comment]
@iamstarkov 	what do you think about using `git` for the FS commands? @kayoub5 has built this functionality for GitHub and GitLab so far - where we support both API as well as git approaches to files - and he could also do the same with Bitbucket. The problem is that if he does it now it will require some refactoring that will then cause conflicts with this branch, so it's up to you whether you think it will save time or require extra time.

FYI here is the list of commands that can be supported via gitFs: https://github.com/renovatebot/renovate/blob/9534e6a9a9587254bad866ca9010834f94dca185/lib/platform/git/storage.js#L18-L30

So basically you could remove all the API equivalents of these and not have to worry about porting them to Bitbucket Server.

**Comment by iamstarkov** [comment]
@rarkins @kayoub5 we are fine with cloud adapter rewrite. in the end of the day it means that only part of our adjustments to BB server will be needed

**Comment by iamstarkov** [comment]
ahaha, bitbucket server allows to delete endpoint only through non-documented api endpoint. 

Also if you want to reset the branch it will throw an exception because the file used to be in that branch already

**Comment by iamstarkov** [comment]
we are going to try gitFs

**Comment by rarkins** [comment]
> we are going to try gitFs

Integrating it is a little tough, although of course you can follow the examples from GitHub and GitLab, which are already migrated.

However if you can wait a little then @kayoub5 	hopefully has time to submit a PR migrating BitBucket Cloud and you could rebase based on that (with some conflicts unfortunately, but  unavoidable).

**Comment by JCB-K** [comment]
@rarkins we setup gitFs for BB server now and it works a charm. This way we avoid migrating a lot of API calls to obscure  BB Server equivalents :). I think we covered most of the happy path, tomorrow we'll focus on error handling + refactoring the server code into its own module.

**Comment by JCB-K** [comment]
@kayoub5 not sure if you started on migrating BB Cloud to gitFs, but we should be able to cover that in this PR too if you haven't started.

**Comment by rarkins** [comment]
Amazing work @JCB-K !

I’m pretty sure @kayoub5 hasn’t started yet, so any work in this PR would be great. 

Does it work using the same token for both git as well as API? For GitLab we found it has a limitation of only supporting git write operations via SSH, which means you need both an SSH key as well as API token. 

I think we could even make both Bitbucket Cloud and Server “gitFs-only” if it simplifies maintenance. I still classified BB support in Renovate as “beta” until now so have no problem deprecating the API-based functions if you think that’s a good idea.

**Comment by JCB-K** [comment]
> Does it work using the same token for both git as well as API?

For now we do git operations on http://username:password@host. BB Server allows you to generate tokens which then can be used as passwords (so "single-use password" would be a more correct term), so I don't think we'll need SSH keys.

> I still classified BB support in Renovate as “beta” until now so have no problem deprecating the API-based functions if you think that’s a good idea.

Aren't all these API methods internal anyway? Or are some of them exposed to users? Don't mind spending some time adopting the functions that aren't available in gitFs to the Server API if they're useful.

**Comment by JCB-K** [comment]
OK cool, we'll remove methods in `lib/platform/bitbucket/index.js` as we see fit then.

**Comment by JCB-K** [comment]
@iamstarkov can you edit the first comment and add the following:

# To Do:

- [ ] Make following methods work for Server, or deprecate if covered by gitFs

## Initialization

- [ ] getRepos
- [ ] cleanRepo
- [x] initRepo
- [x] getRepoStatus
- [ ] getRepoForceRebase
- [x] setBaseBranch

## Search

- [x] getFileList

## Branch

- [x] branchExists
- [x] getAllRenovateBranches
- [x] isBranchStale
- [x] getBranchPr
- [ ] getBranchStatus
- [ ] getBranchStatusCheck
- [ ] setBranchStatus
- [x] deleteBranch
- [ ] mergeBranch
- [ ] getBranchLastCommitTime

## Issue

- [ ] findIssue
- [ ] ensureIssue
- [ ] ensureIssueClosing
- [ ] addAssignees
- [ ] addReviewers
- [ ] deleteLabel

## Comments

- [ ] ensureComment
- [ ] ensureCommentRemoval

## PR

- [x] getPrList
- [x] findPr
- [x] createPr
- [x] getPr
- [ ] getPrFiles
- [ ] updatePr
- [ ] mergePr
- [x] getPrBody

## file

- [x] commitFilesToBranch
- [x] getFile

## Commits

- [x] getCommitMessages

## vulnerability alerts

- [ ] getVulnerabilityAlerts

- [ ] Separate Cloud & Server
  - [ ] create bb-server-got-adapter
- [ ] Ensure Cloud still works
- [ ] implement gitFs for both
- [ ] refactor/clean
- [ ] test coverage for Server
- [ ] implement HTTPS for gitFs
    - [ ] Proper setup for username/password for Server

**Comment by JCB-K** [comment]
@rarkins Could you do something similar as this comment? https://github.com/renovatebot/renovate/pull/1226#issuecomment-347426691 I can't find the specific commit since some stuff has been rebased since then, but it would be very useful to have the same for BB Server. We're planning to separate cloud & server versions today so we can restore the cloud clode to its original version.

**Comment by rarkins** [comment]
@JCB-K I want to make sure I understand:

1. You want me to make edits to this WIP branch/PR
2. You want me to help you understand which functionality is critical and which can be deferred?
3. I'll do it like last time - throw new Error() if it's critical or `// TODO` if it's not?

**Comment by iamstarkov** [comment]
@rarkins 

1. yes, please add a commit with a new file with platform adapter' skeleton and push it to this branch
2. yes, that would be lovely
3. yes, throw for critical functionality and todo comment for optional

**Comment by rarkins** [comment]
@iamstarkov sorry, confused by why it's a new file? 

Also I think this question is relevant: do you now think it's best to have "bitbucket" as a single platform that switches between cloud and server, or should we rename the existing "bitbucket" to "bitbucket-cloud" and create a new "bitbucket-server" ?

**Comment by JCB-K** [comment]
@rarkins we want to create a new new platform now

**Comment by rarkins** [comment]
OK so I'll create a new platform/bitbucket-server/index.js skeleton including all functions?

**Comment by iamstarkov** [comment]
that will work

**Comment by rarkins** [comment]
I had a free 10 minutes so just went through it very fast and pushed it now, hopefully it's mostly accurate. FYI I based it off the gitlab/index.js so you might see some "iid" references, don't be confused.

**Comment by JCB-K** [comment]
@rarkins sweet, we'll take it from here!

**Comment by rarkins** [comment]
IF you have any questions, feel free to @rarkins me on specific lines in that commit and I can hopefully answer from mobile

**Comment by PixnBits** [comment]
Sorry if this is an uninformed question, but which version of the Bitbucket Server API is this being coded to? Asking as I've been using https://docs.atlassian.com/bitbucket-server/rest/5.16.0/bitbucket-rest.html but https://github.com/iamstarkov/renovate/blob/feat/bitbucket-server/test/platform/bitbucket-server/index.spec.js#L33-L35 has a `full_name` key I don't see in https://docs.atlassian.com/bitbucket-server/rest/5.16.0/bitbucket-rest.html#idm46229603858160 .

Looks like this is getting close (woot!) and was looking to help out getting the specs running, but not sure what form of sample data would actually be helpful 😅. Cheers!

**Comment by rarkins** [comment]
If you can, the simplest help right now would be to help get test coverage up - and obviously help verify functions are working as intended in the process. Otherwise testing against a real Bitbucket Server is of course useful.

**Comment by PixnBits** [comment]
I was hoping to use docker to create mock data and possibly later integration test, but the image requires a license to run so those are out.
I'm still curious on if this targeting the [Bitbucket Server 1.0 REST API](https://docs.atlassian.com/bitbucket-server/rest/5.16.0/bitbucket-rest.html) as I can use that document (and a Bitbucket server I have access to) for the mock data. If a different document is the contract targeted that'd be good to know too so I can use that instead. 🙏

**Comment by k2snowman69** [comment]
@PixnBits Looks like it's using the 1.0 api (just pulled the code.

I think my plan for the weekend is going to be copying the data in `test/_fixtures/bitbucket/responses.js` and getting a bitbucket server version of it which should allow for quite a bit of the test code to be copied from one to the other. Hopefully more information by Monday

**Comment by PixnBits** [comment]
Howdy, @k2snowman69! That's what I was expecting myself, but I'm [confused as the existing work doesn't seem to match what's in the API docs?](https://github.com/renovatebot/renovate/pull/2774#issuecomment-444318225) Maybe I've been looking in the wrong spot? 😅🤙

**Comment by k2snowman69** [comment]
I'm using 5.14.1 and I see what you're seeing as well. I'm curious if it's just a copy paste issue from wherever the original was posted from which does use `full_name` (e.g. bit bucket cloud).

I'll add to the top of the responses file the version I'm aligning to for the sample code. Overall if the code needed to branch (e.g. in the next version) we could call `/rest/api/1.0/application-properties` to get the version number and branch depending. I'll leave that up to the owners though :-)

**Comment by rarkins** [comment]
@PixnBits if you're interested in testing against a "real" Bitbucket Server then I'd be happy to buy a license for you to use (https://www.atlassian.com/licensing/bitbucket-server#serverlicenses-1). Please email me so we can set it up.

**Comment by rarkins** [comment]
@PixnBits @k2snowman69 	it's definitely possible that there remain some copy paste errors. I think this describes how this `bitbucket-server` platform evolved:

1. Copy paste from existing `bitbucket` platform, which works for cloud
2. Insert in gitFs support and take out the equivalent API functions (e.g. commit files, list branches, etc).

In both these steps it's quite possible there's some "nonsense" code remaining so feel free to always question it

File: test/platform/bitbucket-server/index.spec.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    const oldGet = api.get;
+    try {
+      api.get = jest.fn().mockImplementation(mockedGet);
+      return await fn();
+    } finally {
+      api.get = oldGet;
+    }
+  }
+  async function mockedGet(path) {
+    let body = responses[URL.parse(path).pathname] || { values: [] };
```
Comment (rarkins) [inline test/platform/bitbucket-server/index.spec.js]: `URL.parse()` returns results in lower case while the fixtures file uses uppercase `SOME` in its response keys

**Comment by rarkins** [comment]
Test coverage is now up to 67% of this PR. Once we hit 100%, we can merge this to master. Can anyone help?

**Comment by iamstarkov** [comment]
@k2snowman69 @PixnBits we used api docs for v5.11.1 , URL is here https://docs.atlassian.com/bitbucket-server/rest/5.11.1/bitbucket-rest.html

because that is the BB version we have in prod. With that said, Feel free to adjust the code for latest version available.

**Comment by k2snowman69** [comment]
> Also, Ricky Patel please sign CLA =) is it you, @k2snowman69?

Not sure what's going on here but I believe I signed the CLA request as k2snowman69 but it doesn't seem to sync up with my name. But yes that is me :-)

**Comment by rarkins** [comment]
Hi all, I'm planning to merge this PR as-is so that it is easier for everyone to test out locally as well as submit PRs against the master branch whenever you find something needing improvement. 

I've been testing it myself with a local Bitbucket Server instance running inside Docker and it onboards, creates and updates PRs, etc.

**Comment by iamstarkov** [comment]
This is amazing news.

@rarkins thank you very much for this

**Pull request merged**